lib/denormalize.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+var _ = require('lodash');
+
+function getLinkedResources(primaryName, linkName, body){
+  var path = primaryName + '.' + linkName;
+  var type = body.links[path] && body.links[path].type;
+  return type && body.linked && body.linked[type];
+}
+
+function rebaseLinks(links, top, next){
+  return _.reduce(links, function(memo, value, key){
+    if (key.indexOf(top) !== 0) return memo;
+    var subtype = links[top + '.' + next].type;
+    var subkey = key.replace(new RegExp(top + '.\\w+'), subtype);
+    if (subkey.indexOf('.') !== -1) memo[subkey] = value;
+    return memo;
+  }, {});
+}
+
+exports._rebaseLinks = rebaseLinks;
+exports._getLinkedResources = getLinkedResources;
+exports.denormalize = function(config, res){
+  var body = res.body;
+  if (config.request.denormalize){
+    denormalizeLayer(body);
+  }
+  function denormalizeLayer(body){
+    var primary = _.without(_.keys(body), 'linked', 'links')[0];
```

Comment:
They way to detect `primary` does not look very reliable. JSON API allows several other keys under root, including `errors` and `meta`.
Is there a way to detect primary key from `config`, for example?